tests/core/test_lightning_optimizer.py

```diff
@@ -0,0 +1,192 @@
+    trainer.fit(model)
+
+    groups = "{'dampening': 0, 'initial_lr': 0.1, 'lr': 0.01, 'momentum': 0, 'nesterov': False, 'weight_decay': 0}"
+    expected = f"LightningSGD(groups=[{groups}])"
+    assert trainer.optimizers[0].__repr__() == expected
+
+
+def test_lightning_optimizer_from_user(tmpdir):
+    """
+    Test that the user can use our LightningOptimizer. Not recommended.
+    """
+
+    class TestModel(BoringModel):
+        def configure_optimizers(self):
+            optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+            optimizer = LightningOptimizer(optimizer)
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
+            return [optimizer], [lr_scheduler]
+    model = TestModel()
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
+        limit_train_batches=1,
+        limit_val_batches=1,
+        max_epochs=1,
+        weights_summary=None,
+        enable_pl_optimizer=True)
+    trainer.fit(model)
+
+    groups = "{'amsgrad': False, 'betas': (0.9, 0.999), 'eps': 1e-08, 'initial_lr': 0.1, 'lr': 0.01, 'weight_decay': 0}"
+    expected = f"LightningAdam(groups=[{groups}])"
+    assert trainer.optimizers[0].__repr__() == expected
+
+
+@patch("torch.optim.Adam.step")
+@patch("torch.optim.SGD.step")
+def test_lightning_optimizer_manual_optimization(mock_sgd_step, mock_adam_step, tmpdir):
+    """
+    Test that the user can use our LightningOptimizer. Not recommended for now.
+    """
+    class TestModel(BoringModel):
+
+        def training_step(self, batch, batch_idx, optimizer_idx=None):
+            (opt_1, opt_2) = self.optimizers()
+            assert isinstance(opt_1, LightningOptimizer)
+            assert isinstance(opt_2, LightningOptimizer)
+
+            output = self.layer(batch)
+            loss_1 = self.loss(batch, output)
+            self.manual_backward(loss_1, opt_1)
+            opt_1.step(idx="1")
+
+            def closure():
+                output = self.layer(batch)
+                loss_2 = self.loss(batch, output)
+                self.manual_backward(loss_2, opt_2)
+            opt_2.step(closure=closure, idx="2")
+
+        def configure_optimizers(self):
+            optimizer_1 = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            optimizer_2 = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+            optimizer_1 = LightningOptimizer(optimizer_1, 4)
+
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer_1, step_size=1)
+            return [optimizer_1, optimizer_2], [lr_scheduler]
+
+    model = TestModel()
+    model.training_step_end = None
+    model.training_epoch_end = None
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
+        limit_train_batches=8,
+        limit_val_batches=1,
+        max_epochs=1,
+        weights_summary=None,
+        automatic_optimization=False,
+        enable_pl_optimizer=True)
+    trainer.fit(model)
+
+    assert len(mock_sgd_step.mock_calls) == 2
```

Comment:
does it have something like nb calls already?

Reply:
Not sure, I will look into it and use it next time.